symengine/ruby/ext/symengine/symengine_macros.c

```diff
@@ -0,0 +1,51 @@
+#include "symengine_macros.h"
+#include "symengine.h"
+
+void sympify(VALUE operand2, basic_struct **cbasic_operand2) {
+    basic_struct *num_basic, *den_basic;
+    VALUE new_operand2, num, den;
+
+    switch(TYPE(operand2)) {
+        case T_FIXNUM:
+        case T_BIGNUM:
+            new_operand2 = cbasic_alloc(c_integer);
+            Data_Get_Struct(new_operand2, basic_struct, *cbasic_operand2);
+            GET_SYMINTFROMVAL(operand2, *cbasic_operand2);
+            break;
+
+        case T_RATIONAL:
+            num = rb_funcall(operand2, rb_intern("numerator"), 0, NULL);
+            den = rb_funcall(operand2, rb_intern("denominator"), 0, NULL);
+
+            num_basic = basic_new_heap();
+            den_basic = basic_new_heap();
+
+            GET_SYMINTFROMVAL(num, num_basic);
+            GET_SYMINTFROMVAL(den, den_basic);
+
+            new_operand2 = cbasic_alloc(c_rational);
```

Comment:
Exactly. The only thing you need to make sure is to do a copy of the RCP
(which increments refcount) from Ruby. That way when the Ruby object goes
out of scope, the underlying SymEngine object doesn't get deallocated if
you still use it elsewhere.

Sent from my mobile phone.
On Aug 4, 2015 10:25 AM, "Abinash Meher" <notifications@github.com> wrote:

> In symengine/ruby/ext/symengine/symengine_macros.c
> <https://github.com/sympy/symengine/pull/487#discussion_r36209075>:
>
> > +            new_operand2 = cbasic_alloc(c_integer);
> > +            Data_Get_Struct(new_operand2, basic_struct, *cbasic_operand2);
> > +            GET_SYMINTFROMVAL(operand2, *cbasic_operand2);
> > +            break;
> > +
> > +        case T_RATIONAL:
> > +            num = rb_funcall(operand2, rb_intern("numerator"), 0, NULL);
> > +            den = rb_funcall(operand2, rb_intern("denominator"), 0, NULL);
> > +
> > +            num_basic = basic_new_heap();
> > +            den_basic = basic_new_heap();
> > +
> > +            GET_SYMINTFROMVAL(num, num_basic);
> > +            GET_SYMINTFROMVAL(den, den_basic);
> > +
> > +            new_operand2 = cbasic_alloc(c_rational);
>
> Ah! I get it now. Yes, sympify is meant to return a SymEngine object. I
> wonder why this didn't come to me earlier. We can always extract the
> information from Ruby objects and return a SymEngine object. And after
> sympify has returned, we can wrap it, if needed. Thanks @certik
> <https://github.com/certik>!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/symengine/pull/487/files#r36209075>.
>